### PR TITLE
concept of a fptas for the knapsack problem

### PR DESCRIPTION
**Pull request by Mathemalsky**

### CODE REVIEW

**Comment by Mathemalsky** [comment]
Possible improvements:
- The performance gain due to the rounding is still random because of floating point arithmetic.
- The algorithm could also return a a fortiori guarantee.
- [x] We could use `std::pair` instead of `std::tuple`

File: src/AAL/AAL_knapsack.cpp [hunk 1]
```diff
@@ -1,6 +1,8 @@
 #include "AAL_knapsack.h"
 
 #include <algorithm>
+#include <cmath>
+#include <exception>
```
Comment (MilchRatchet) [inline src/AAL/AAL_knapsack.cpp]:
```suggestion
#include <stdexcept>
```
`std::runtime_error` is part of `<stdexcept>`. The standard `<exception>` header does not define `std::runtime_error`.

File: TEST/TEST_knapsack.cpp [hunk 2]
```diff
@@ -32,19 +34,34 @@ void TEST_KNAPSACK_EXACT() {
...
 
   unsigned int* subset = (unsigned int*) malloc(3 * sizeof(unsigned int));
   subset[0]            = 0;
   subset[1]            = 5;
   subset[2]            = 7;
 
-  print_check(
-    std::abs(result.objectiveValue - 13.1f) < 0.00001f && compareAryUInt(result.items, result.numberOfItems, subset, 3),
-    "AAL_KNAPSACK_EXACT");
+  if (testset.count("KSE") != 0) {
```
Comment (MilchRatchet) [inline TEST/TEST_knapsack.cpp]: The knapsack test does not run if for example `FULL` or `AAL` are given. I think the best way is to set the flags earlier in the function and then execute based on if the respective flag is set.

File: TEST/TEST_knapsack.cpp [hunk 3]
```diff
@@ -32,19 +34,34 @@ void TEST_KNAPSACK_EXACT() {
...
+
+    print_check(
+      std::abs(result.objectiveValue - 13.1f) < 0.00001f
+        && compareAryUInt(result.items, result.numberOfItems, subset, 3),
+      "AAL_KNAPSACK_EXACT");
+
+    free(result.items);
+  }
+  if (testset.count("KSF") != 0) {
+    const float eps           = 0.1;
```
Comment (MilchRatchet) [inline TEST/TEST_knapsack.cpp:54]:
```suggestion
    const float eps           = 0.1f;
```

File: src/AAL/AAL_knapsack.cpp [hunk 4]
```diff
@@ -84,3 +86,35 @@ AAL_knapsackResult AAL_knapsack_exact(AAL_knapsackInstance instance) {
...
+  for (unsigned int i = 0; i < instance.numberOfItems; ++i) {
+    roundedItemValues[i] = std::floor(instance.itemValues[i] / mu) * mu;
+  }
+  AAL_knapsackInstance roundedInstance{
+    instance.itemSizes, roundedItemValues, instance.numberOfItems, instance.capacity};
+  // invoke the exact knapsack on the rounded data
+  AAL_knapsackResult result = AAL_knapsack_exact(roundedInstance);
+  free(roundedItemValues);
+  // retranslate the solution on the rounded data to the original data
+  float objectiveValue = 0;
```
Comment (MilchRatchet) [inline src/AAL/AAL_knapsack.cpp]:
```suggestion
  float objectiveValue = 0.0f;
```

**Review by MilchRatchet — CHANGES_REQUESTED** [review]
I have noted some small things that need changes.

File: TEST/TEST_knapsack.cpp (hunk 2 above)
Comment (MilchRatchet) [reply TEST/TEST_knapsack.cpp]: The highlighted code by Github is a bit misleading, I simply mean that the if does not check for all cases.

**Review by MilchRatchet — APPROVED** [review]

**Pull request merged**